test/system/helpers.network.bash

```diff
@@ -140,6 +142,32 @@ function random_rfc1918_subnet() {
     die "Could not find a random not-in-use rfc1918 subnet"
 }
 
+# subnet_in_use() - true if subnet already routed on host
+function subnet_in_use() {
+    # Octets as /24 in binary
+    local binaddr=$(perl -e 'printf "%08b%08b%08b", @ARGV' "$@")
+
+    # This runs 'ip route list', converts x.x.x.x/n to its binary prefix,
+    # then checks if our desired subnet matches that prefix (i.e. is in
+    # that range). Existing routes with size greater than 24 are
+    # normalized to /24 because that's the granularity of our
+    # random_rfc1918_subnet code.
+    #
+    # Contrived examples:
+    #    127.0.0.0/1   -> 0
+    #    128.0.0.0/1   -> 1
+    #    10.0.0.0/8    -> 00001010
+    #
+    # I'm so sorry for the ugliness.
+    while read prefix; do
+        if [[ "$binaddr" =~ ^$prefix ]]; then
+            return 0
+        fi
+    done < <(ip route list | perl -ne 'if (m!^([\d.]+)/(\d+)!) { @ip = split(/\./, $1); $bits = $2; $bits = 24 if $bits > 24; $mask = (2 ** $bits - 1) << (32 - $bits); $ip = unpack("N",pack("C4",@ip)); $net = sprintf("%032b", ($ip & $mask)); printf "%.*s\n", $bits, $net }')
```

Comment:
it is extremely hard to follow that line, can you somehow pretty print it instead of smashing into one gigantic one liner?

Also in case it helps you parsing the output you could run `ip -j route` which gives you json so you wouldn't need the more complex regex to filter, i.e. `ip -j route | jq -r .[].dst` would look much simpler to me to filter the destination addresses, of course you still need to filter out "default" from that list.

Reply:
Rewritten and repushed. The rewrite let me see unnecessary code that I have now excised, yay.

Tests running simultaneously now here and in #17831 and #20161.